lib/SPIRV/libSPIRV/SPIRVValue.h

```diff
@@ -98,8 +98,35 @@ class SPIRVValue : public SPIRVEntry {
 
   void setAlignment(SPIRVWord);
   void setVolatile(bool IsVolatile);
-  void setNoSignedWrap(bool HasNoSignedWrap);
-  void setNoUnsignedWrap(bool HasNoUnsignedWrap);
+
+  template <spv::Decoration NoIntegerWrapDecoration>
+  void setNoIntegerDecorationWrap(bool HasNoIntegerWrap) {
+    if (!HasNoIntegerWrap) {
+      eraseDecorate(NoIntegerWrapDecoration);
+      return;
+    }
+    // NoSignedWrap and NoUnsignedWrap decorations are available only if it is
+    // allowed to use SPIR-V 1.4 or if SPV_KHR_no_integer_wrap_decoration
+    // extension is enabled
+    if (Module->isAllowedToUseVersion(VersionNumber::SPIRV_1_4)) {
+      Module->setMinSPIRVVersion(
+          static_cast<SPIRVWord>(VersionNumber::SPIRV_1_4));
+      addDecorate(new SPIRVDecorate(NoIntegerWrapDecoration, this));
+      SPIRVDBG(spvdbgs() << "Set "
+                         << (NoIntegerWrapDecoration == DecorationNoSignedWrap)
+                   ? "nsw"
+                   : "nuw"
+                         << " for obj " << Id << "\n")
+    } else if (Module->isAllowedToUseExtension(
+                   ExtensionID::SPV_KHR_no_integer_wrap_decoration)) {
+      Module->addExtension(ExtensionID::SPV_KHR_no_integer_wrap_decoration);
+      addDecorate(new SPIRVDecorate(NoIntegerWrapDecoration, this));
+      SPIRVDBG(spvdbgs() << "Set nsw for obj " << Id << "\n")
+    } else {
+      SPIRVDBG(spvdbgs() << "Skip setting nsw for obj " << Id << "\n")
```

Comment:
Here as well. Maybe better to make a variable for that value?

Reply:
Applied